File: 1/exercise1/src/task1.test.c
/// @file
/// @brief This program creates "task1_output_c.txt" and writes details about
/// when the program was compiled, which C standard was used, and which source
/// file contains the main function. It then asserts that the compiler supports
/// at least the C11 standard and prints a confirmation message if all checks
/// pass.

/// @todo Ensure you can compile and run this C program in your environment.
/// compile & run command from exercise1/:
///   mkdir -p build output && gcc -std=c11 src/task1.test.c -o build/task1c && ./build/task1c

#include <assert.h> // assert
#include <stdio.h>  // FILE
#include <stdlib.h> // EXIT_FAILURE|EXIT_SUCCESS

static const char* compiler =
#if defined(__clang__)
    "Clang " __clang_version__;
#elif defined(__GNUC__)
    "GCC " __VERSION__;
#elif defined(_MSC_VER)
    "MSVC";
#else
    "Unknown compiler";
#endif

int main(void)
{

    FILE* stream = fopen("output/task1_output_c.txt", "w");
    if (stream == NULL)
        return EXIT_FAILURE;

    fprintf(stream, "This compilation unit ( %s ) was compiled\n", __FILE__);
    fprintf(stream, "  - on %s at %s,\n", __DATE__, __TIME__);
    fprintf(stream, "  - using the %s compiler,\n", compiler);
    fprintf(stream, "  - the C language standard %li.\n", __STDC_VERSION__);

    fclose(stream);

    assert(__STDC_VERSION__ >= 201112L);

    printf("task1.test.c: all asserts passed\n");

    return EXIT_SUCCESS;
}

File: 1/exercise1/src/task1.test.py
#!/usr/bin/env python3
"""
This script writes runtime environment details to "task1_output_python.txt"
including the script filename, execution date and time, Python implementation
and versions for NumPy and Matplotlib.

TODO:
  Ensure you can run this Python script in your environment
  run command from exercise1/:
  mkdir -p build output && python src/task1.test.py
"""

import platform
import numpy
import matplotlib
from datetime import datetime
from pathlib import Path

with open("output/task1_output_python.txt", "w") as f:
    f.write(
        f"""This script ({Path(__file__).name}) was executed
  - on {datetime.now()},
  - using {platform.python_implementation()} {platform.python_version()},
  - with {numpy.__name__} {numpy.__version__}, and
  - with {matplotlib.__name__} {matplotlib.__version__}."""
    )

print("task1.test.py: all asserts passed")
